win32ss/user/winsrv/usersrv/harderror.c

```diff
@@ -439,6 +439,58 @@ UserpFormatMessages(
     return Status;
 }
 
+static void
+UserpShowInformationBalloon(
+    PWSTR Text,
+    PWSTR Caption,
+    ULONG ValidResponseOptions,
+    ULONG Status)
+{
+    HWND hShellTrayWnd;
+    COPYDATASTRUCT CopyData;
+    PBALLOON_HARD_ERROR_DATA pdata;
+    DWORD dwSize, cchText, cchCaption;
+    PWCHAR pText, pCaption;
+
+    hShellTrayWnd = FindWindowW(L"Shell_TrayWnd", NULL);
+    if (!hShellTrayWnd)
+    {
+        DPRINT1("Failed to find Shell_TrayWnd\n");
+        return;
+    }
+
+    cchText = wcslen(Text);
+    cchCaption = wcslen(Caption);
+
+    dwSize = sizeof(BALLOON_HARD_ERROR_DATA);
+    dwSize += (cchText + 1) * sizeof(WCHAR);
+    dwSize += (cchCaption + 1) * sizeof(WCHAR);
+
+    pdata = RtlAllocateHeap(RtlGetProcessHeap(), HEAP_ZERO_MEMORY, dwSize);
+    if (!pdata)
+    {
+        DPRINT1("Failed to allocate balloon package\n");
+        return;
+    }
+
+    pdata->cbHeaderSize = sizeof(BALLOON_HARD_ERROR_DATA);
+    pdata->Status = Status;
+    pdata->dwResponseType = ValidResponseOptions;
+    pdata->TitleOffset = pdata->cbHeaderSize;
+    pdata->MessageOffset = pdata->TitleOffset;
+    pdata->MessageOffset += (cchCaption + 1) * sizeof(WCHAR);
+    pCaption = (PWCHAR)((ULONG_PTR)pdata + pdata->TitleOffset);
+    pText = (PWCHAR)((ULONG_PTR)pdata + pdata->MessageOffset);
+    wcscpy(pCaption, Caption);
+    wcscpy(pText, Text);
+
+    CopyData.dwData = BALLOON_DATA;
+    CopyData.cbData = dwSize;
+    CopyData.lpData = pdata;
+
+    SendMessageW(hShellTrayWnd, WM_COPYDATA, 0, (LPARAM)&CopyData);
```

Comment:
This must timeout, otherwise if for some reason the explorer hangs in there, then all the csrss (or at least all the hard error part) would suddently hang forever.

Reply:
right

Reply:
I remember seeing csrss showing a normal message box after a timeout. I think this happened either when explorer was not running or it was suspended.

Reply:
That's correct, if the SendMessage(Timeout) fails, because the explorer is hanging and cannot treat the balloon notification, then csrss calls instead MessageBoxTimeout.